### ISSUE TITLE
DiagnostSource async and parenting

### ISSUE THREAD

**Issue by Wraith2** [dotnet/corefx#37966]
I'm writing a program that and want to be able to monitor non-intrusively in production. The suggested way to do this was EventSource in the past but is now DiagnosticSource which should work for both in and out of process mornitoring.

I'm receiving messages from client systems and once validated I want to schedule a new task to deal with the contents. Once I've received the message the receive task is done and the processing step is logically distinct so I create a disconnected task to deal with it. When using diagnostic source even though the new task is disconnected it inherits the Activity.Current async local from the calling context and appears in output as connected.

I've put together a sample project to investigate and show this [here](https://github.com/Wraith2/DiagnosticsPlayground)

In the standard way of running you get output like [this](https://github.com/Wraith2/DiagnosticsPlayground/blob/4db70b3206f573ac8722b7e5f764957cd81a7da7/sample_connected.txt#L5-L11):
```
|40d2d111-4f7df7289f0995b0.1.1.        ReceiveEvent.Take     [ rid=1 ] { { rid = 1 } }
|40d2d111-4f7df7289f0995b0.1.1.1.        ProcessEvent.Start  [ eid=1, rid=1 ]
|40d2d111-4f7df7289f0995b0.1.1.1.        ProcessEvent.Stop   [ eid=1, rid=1 ]
|40d2d111-4f7df7289f0995b0.1.1.        ReceiveEvent.Stop     [ rid=1 ]
|40d2d111-4f7df7289f0995b0.1.1.1.1.         InternalProcess.Start  [ eid=1, eid=1, rid=1 ]
|40d2d111-4f7df7289f0995b0.1.1.1.1.         InternalProcess.Work  [ eid=1, eid=1, rid=1 ] { { result = 2 } }
|40d2d111-4f7df7289f0995b0.1.1.1.1.         InternalProcess.Stop  [ eid=1, eid=1, rid=1 ]
```

Where the ordering is correct because ReceiveEvent finishes before InternalProcess starts because of the disconnected task but the event id's and path are confusingly linked. 

Currently the only way I can see to change this is to suppress execution context flow while creating the task, which changes the output to be like [this](https://github.com/Wraith2/DiagnosticsPlayground/blob/4db70b3206f573ac8722b7e5f764957cd81a7da7/sample_disconnected.txt#L3-L11):
```
|bb588b1a-43bb65abf369eeec.1.1.        ReceiveEvent.Start    [ rid=1 ]
|bb588b1a-43bb65abf369eeec.1.1.        ReceiveEvent.Take     [ rid=1 ] { { rid = 1 } }
|bb588b1a-43bb65abf369eeec.1.1.1.        ProcessEvent.Start  [ eid=1, rid=1 ]
|bb588b1a-43bb65abf369eeec.1.1.1.        ProcessEvent.Stop   [ eid=1, rid=1 ]
|bb588b1a-43bb65abf369eeec.1.1.        ReceiveEvent.Stop     [ rid=1 ]
|bb588b1b-43bb65abf369eeec.        InternalProcess.Start     [ eid=1 ]
|bb588b1b-43bb65abf369eeec.        InternalProcess.Work      [ eid=1 ] { { result = 2 } }
|bb588b1b-43bb65abf369eeec.        InternalProcess.Stop      [ eid=1 ]
```
Which shows a (correctly in my opinion) the ReceiveEvent occurring and finishing followed by the InternalProcess logically disconnected from it. It might be nice to reparent the InternalProcess activity from somewhere but I would need to set that up manually.

Supressing the execution context flow feels heavy handed. There are other things which may be using async locals (impersonation for example) that I might want to keep but at the moment I don't have the option to do so. Am I misunderstanding something about the api or how it should be used? Should there be (or is there) a simple way to disconnect a new activity from the current parent without supressing execution context flow?

Are there any examples of getting diagnostic events out of process through ETW? The way to do this seems to be in [DiagnosticSourceEventSource](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs) but neither the extensive comment or the tests give me enough information as a diagnostics beginner to work out what needs to be done and I can't find any examples anywhere that use diagnostics from out of process. 

Ideally I want to be able to connect to the machine with the software already running and fire up prefview to gather a profile trace with the diagnostics events only being enabled when the trace is being run, I think this is an intended use case.

/cc @noahfalk, @tommcdon, @Anipik

**Comment by noahfalk** [comment]
> The suggested way to do this was EventSource in the past but is now DiagnosticSource which should work for both in and out of process mornitoring.

Both techniques are usable in-proc, the primary difference is that EventSource requires all the logging data passing through it is serializable whereas DiagnosticsSource allows passing object references directly with no serialization. If serialization is OK EventSource + EventListener should work well. If you need to pass around object references DiagnosticsSource + DiagnosticsListener fills that role.

> Supressing the execution context flow feels heavy handed. There are other things which may be using async locals (impersonation for example) that I might want to keep but at the moment I don't have the option to do so. Am I misunderstanding something about the api or how it should be used?

It feels a little counter-intuitive to me that for logging purposes you want to treat a task as not being connected to the parent, but you'd like all the other state by default to be connected to the parent. The options that feel less surprising to me if I saw them in code would be:
1) all-or-nothing (what you get by suppressing execution context)
2) Opt-in certain async state that you need explicitly - you would need to do something to capture the state you care about before spawning the task and then recreate that same state within the new task's thread.

I'm not claiming there is something inherently wrong in selectively disabling Activity flow, but it does at least feel atypical. It would be good in your own head at least to be clear on the rationale that makes logging a special case that should be treated differently than all other async state.

> Should there be (or is there) a simple way to disconnect a new activity from the current parent without supressing execution context flow?

In terms of disconnecting the Activity from its async Parent, I think you could also use [Activity.SetParentId(string)](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L243) prior to calling Start(). This will cause the Activity to root itself using the parent Id you picked rather than the ambient activity. One warning here is that if your thread's current activity is A and then you use this technique to set a new root activity R, after calling R.Stop() the thread will have CurrentActivity=null, not A. 

> Are there any examples of getting diagnostic events out of process through ETW? The way to do this seems to be in DiagnosticSourceEventSource but neither the extensive comment or the tests give me enough information as a diagnostics beginner to work out what needs to be done and I can't find any examples anywhere that use diagnostics from out of process.

Apologies, I'm not aware of a good example or docs beyond the comment/tests that you found. I think DiagnosticsSourceEventSource is a bit of a niche area of the product that has only gotten limited use. I opened issue #24873 to create a better example and I might be able to answer some questions but I'll warn I've never had the opportunity to really dig into it in the past. Another alternative is that it is perfectly legal to use EventSource and DiagnosticsSource at the same time. There are a lot more examples/docs for EventSource and how to use it with ETW consuming tools like PerfView.

HTH,
 -Noah

**Comment by Wraith2** [comment]
I intend for my data to be serializable and not to pass around complex types so I could use EventSource but having looked around I'd seen recommendations that diagnostics activities were the way to do this since it can be bridged backwards. I could rework on event source but that required writing an event source class for each use case and that feels like a step backwards in usability.

> It feels a little counter-intuitive to me that for logging purposes you want to treat a task as not being connected to the parent

I'm working with a protocol which decouples the receipt of a message from the processing of that message. The client sends a message to the server which stores it in a database and then returns a receipt confirmation to the client. The message from the database is picked up elsewhere and processed. In the example traces above you can see this as rid (receipt id) and eid (event id), a receipt happens and generates an event and then that event is processed independently to the receipt. So it's more like a one-way call than a call-reponse that you'd see in aspnet. 

Rather than polling the database on a dedicated loops I queue up a task to process the received message. The receipt call uses Task.Factory.StartNew to create a disconnected task but when it does that new task still inherits the async local current activity. If I were using threads it'd be a producer consumer model but I didn't see any need to have that when I can just start a task and achieve the same thing.

So really the question is whether it should be easier to create a disconnected task and have a way to say that it's also disconnected from the current activity.. The current api supports awaiting/promisary tasks well but makes background tasks confusing imo.

**Comment by noahfalk** [comment]
I think its reasonable to have a scenario with a decoupled task, I was more surprised that you are only trying to apply the decoupling to the activity tracing and not to any other aspect of the async state. 

> So really the question is whether it should be easier to create a disconnected task and have a way to say that it's also disconnected from the current activity.

ExecutionContext.SuppressFlow() will certainly suppress it with a single line of code. I think this appears harder because you are trying to single out Activity and preserve all other async state. This feels like an atypical goal to me.

**Comment by Wraith2** [comment]
In this case I have no other execution state I want to keep so I can use the suppression. It feels like an incorrectly sized hammer for the problem and knowing that it will work requires some pretty advanced knowledge because ExecutionContext isn't something most people will/should ever be near. Perhaps the better option is to set Activity.Current to null in the disconnected task. 

This might not need a change but it's worth considering the usability and perhaps some documentation.
